[samples] (Power Automate) Convert CSV files to workbooks (#373)
OfficeDev/office-scripts-docs: docs/resources/samples/convert-csv.md

---
title: 'Convert CSV files to Excel workbooks'
description: 'Learn how to use Office Scripts and Power Automate to create .xlsx files from .csv files.'
ms.date: 07/19/2021
localization_priority: Normal
---

# Convert CSV files to Excel workbooks

Many services export data as comma-separated value (CSV) files. This solution automates the process of converting those CSV files to Excel workbooks in the .xlsx file format. It uses a [Power Automate](https://flow.microsoft.com) flow to find .csv files in a OneDrive folder and an Office Script to copy the data from the .csv file into a new Excel workbook.

## Solution

1. Store the .csv files and a blank "Template" .xlsx file in a OneDrive folder.
1. Create an Office Script to parse the CSV data into a range.
1. Create a Power Automate flow to read the .csv files and pass their contents to the script.

## Sample files

> Download <a href="https://github.com/OfficeDev/office-scripts-docs/blob/master/docs/resources/samples/convert-csv-example.zip?raw=true">convert-csv-example.zip</a> to get the template .xlsx file and two sample .csv files. Extract the files into a folder in your OneDrive. This sample assumes the folder is named "output".

Reviewer: This link doesn't work for me – it goes to a GitHub 404 page. Am I doing something wrong?

Author: No, links to raw files need to include a branch. I forgot about this quirk and should have specified to use the file in the PR instead of the link.

## Sample code: Insert comma-separated values into a workbook

```TypeScript
function main(workbook: ExcelScript.Workbook, csv: string) {
  /* Convert the CSV data into a 2D array. */
  // Trim the trailing new line.
  csv = csv.trim();

  // Split each line into a row.
  let rows = csv.split("\r\n");
  let data : string[][] = [];
  rows.forEach((value) => {
    /*
     * For each row, match the comma-separated sections.
     * For more information on how to use regular expressions to parse CSV files,
     * see this Stack Overflow post: https://stackoverflow.com/a/48806378/9227753
     */
    data.push(value.match(/(?:,|\n|^)("(?:(?:"")*[^"]*)*"|[^",\n]*|(?:\n|$))/g));
  });

  // Put the data in the worksheet.
  let sheet = workbook.getWorksheet("Sheet1");
  let range = sheet.getRangeByIndexes(0, 0, data.length, data[0].length);
  range.setValues(data);

  // Add any formatting or table creation that you want.
}
```

## Power Automate flow: Create new .xlsx files

1. Sign into [Power Automate](https://flow.microsoft.com) and create a new **Scheduled cloud flow**.
1. Set the flow to **Repeat every** "1" "Day" and select **Create**.

> 1. Get the template Excel file. This is the basis for all the converted .csv files. Add a **New step** that uses the **OneDrive for Business** connector and the **Get file content** action. Provide the file path to the "Template.xlsx" file.

Author: I'm not sure I understand the concern. Selecting the OneDrive for Business connector should bring up all of that connector's actions. Is the recommendation to flip the order because the text appears with "Get file content" above "OneDrive for Business" in the action?

> 1. Get the template Excel file. This is the basis for all the converted .csv files. Add a **New step** that uses the **OneDrive for Business** connector and the **Get file content** action. Provide the file path to the "Template.xlsx" file.

Reviewer: My main point is that the current directions are ambiguous and don't define a clear path to follow. It's not clear whether the user should search the Actions list or the Connectors list. It's possible to do either, but it's confusing that the tutorial doesn't provide guidance about which list to search or that either can be used (confusing anecdotally — this confused me, a novice Power Automate user). 

I think the remedy is to give a clear path. Tell the user to search the Actions list, or tell the user to search the Connectors list. I think the Actions path is easier, because it shows that there's also a SharePoint Action with the same name that we DON'T want to choose (which clarifies an underlying ambiguity).

A different remedy could be to clarify that the user can search either the Actions or the Connectors tab, defining the path more like "choose your own adventure, but they all end up at the same place".

    1. **File**: /output/Template.xlsx

> 1. Rename the **Get file content** step by going to the **...** menu and selecting the **Rename** option. Change the step name to "Get Excel template".

Reviewer: Dialog might not be the right term, but **...** menu confused me, so I suggest adding a description of the menu location. 
```suggestion
1. Rename the **Get file content** step by going to the **...** menu in the upper right corner of the dialog and selecting the **Rename** option. Change the step name to "Get Excel template".
```

Author: Good call. How does this sound?

Rename the **Get file content** step by going the **...** menu of that step (in the upper right corner of the connector) and selecting the **Rename** option. Change the step name to "Get Excel template".

Reviewer: Sounds good to me!

     :::image type="content" source="../../images/convert-csv-flow-1.png" alt-text="The completed OneDrive for Business connector in Power Automate, renamed to be Get Excel template.":::
1. Get all the files in the "output" folder. Add a **New step** that uses the **OneDrive for Business** connector and the **List files in folder** action. Provide the folder path that contains the .csv files.
    * **Folder**: /output

    :::image type="content" source="../../images/convert-csv-flow-2.png" alt-text="The completed OneDrive for Business connector in Power Automate.":::

> 1. Add a condition so that the flow only operates on .csv files. Add a **New step** that is the **Condition** control. Use the following values for the control. Note that when you select dynamic content with multiple results, an **Apply to each** control surrounds the **Condition**.

Author: Judging by the [PA docs](https://docs.microsoft.com/en-us/power-automate/add-condition), conditions don't seem to be actions. The term "action" seemed to be reserved for an option from a service controller.

> 1. Add a condition so that the flow only operates on .csv files. Add a **New step** that is the **Condition** control. Use the following values for the control. Note that when you select dynamic content with multiple results, an **Apply to each** control surrounds the **Condition**.

Reviewer: It's under the Action tab. Maybe something like "uses the Condition control under the Action tab".

>     * **Choose a value**: *Name* (dynamic content from **List files in folder**)

Author: You should only be selecting *Name* once. Are you talking about the "Apply to each" segment that pops up around the condition? I'll try to make it clear this happens during this step.

    * **ends with** (from the dropdown list)
    * **Choose a value**: .csv

    :::image type="content" source="../../images/convert-csv-flow-3.png" alt-text="The completed Condition control with the Apply to each control around it.":::
1. The rest of the flow is under the **If yes** section, since we only want to act on .csv files. Get an individual .csv file by adding a **New step** that uses the **OneDrive for Business** connector and the **Get file content** action. Use the **Id** from the dynamic content from **List files in folder**.
    * **File**: *Id* (dynamic content from the **List files in folder** step)
1. Rename the new **Get file content** step to "Get .csv file". This helps distinguish this file from the Excel template.

> 1. Make the new .xlsx file, using the Excel template as the base content. Add a **New step** that uses the **OneDrive for Business** connector and the **Create file** action. Use the following values.

Reviewer: As discussed offline, we are seeing different versions of Power Automate. My version only has "Add an action" as an option inside If yes. If yours says "New step", maybe we should make this direction more ambiguous? 
```suggestion
1. Make the new .xlsx file, using the Excel template as the base content. Choose **Add an action** and set it to use the **OneDrive for Business** connector and the **Create file** action. Use the following values.
```

    * **Folder Path**: /output
    * **File Name**: *Name without extension*.xlsx (choose the *Name without extension* dynamic content from the **List files in folder** and manually type ".xlsx" after it)
    * **File Content**: *File content* (dynamic content from **Get Excel template**)

     :::image type="content" source="../../images/convert-csv-flow-4.png" alt-text="The Get .csv file and Create file steps of the Power Automate flow.":::
1. Run the script to copy data into the new workbook. Add the **Excel Online (Business)** connector with the **Run script** action. Use the following values for the action.
    * **Location**: OneDrive for Business
    * **Document Library**: OneDrive
    * **File**: *Id* (dynamic content from **Create file**)

>     * **Script**: Convert CSV

Reviewer: I got to this point without ever creating an Office Script called "Convert CSV". Should I have done that before the Power Automate flow?

Author: That should be covered in the "Solution" section. This is the pattern used in other PA/OS samples. Happy to make a change across the board in a separate PR.

Reviewer: Cross-reference Excel files with Power Automate has a line that says "Add the following scripts to try the sample yourself!" — I think adding something as simple as that to this tutorial would prompt the script creation step that I missed.

1. Rename the new **Get file content** step to "Get .csv file". This helps distinguish this file from the Excel template.
1. Make the new .xlsx file, using the Excel template as the base content. Add a **New step** that uses the **OneDrive for Business** connector and the **Create file** action. Use the following values.
    * **Folder Path**: /output
    * **File Name**: *Name without extension*.xlsx (choose the *Name without extension* dynamic content from the **List files in folder** and manually type ".xlsx" after it)
    * **File Content**: *File content* (dynamic content from **Get Excel template**)

     :::image type="content" source="../../images/convert-csv-flow-4.png" alt-text="The Get .csv file and Create file steps of the Power Automate flow.":::
1. Run the script to copy data into the new workbook. Add the **Excel Online (Business)** connector with the **Run script** action. Use the following values for the action.
    * **Location**: OneDrive for Business
    * **Document Library**: OneDrive
    * **File**: *Id* (dynamic content from **Create file**)
    * **Script**: Convert CSV
    * **csv**: *File content* (dynamic content from **Get .csv file**)

> 

Reviewer: I finished the tutorial and am not sure how to test it. When I try to run a manual test, I get an error. :(

Steps 6 and 7 of the [Cross-reference worksheets tutorial](https://docs.microsoft.com/en-us/office/dev/scripts/resources/samples/excel-cross-reference) direct the user to save the flow and then describe how to test it. Is that something we can include for this tutorial?

Author: Added those steps. Can I see the error you're getting?

Reviewer: (Sent error screenshot offline)
